### ISSUE TITLE
Nix 2.21 fails to build flakes when using `--store`

### ISSUE THREAD

**Issue by ipetkov** [NixOS/nix#10267]
**Describe the bug**

I have [a regression test](https://github.com/ipetkov/crane/blob/10484f86201bb94bd61ecc5335b1496794fedb78/extra-tests/alt-store/test.sh) for a project of mine which builds a flake using a different store path via `--store`. It suddenly started failing with the 2.21 release, but works with 2.20 just fine. 

**Steps To Reproduce**

```sh
git clone https://github.com/ipetkov/crane.git
cd crane/extra-tests/alt-store
```
```sh
# Succeeds
nix run nixpkgs#legacyPackages.x86_64-linux.nixVersions.nix_2_20 \
  -- build .#default \
  --no-write-lock-file \
  --override-input crane ../.. \
  --store $(pwd)/alt-store
```
```sh
# Clean up
sudo rm -rf ./alt-store

# Fails
nix run nixpkgs#legacyPackages.x86_64-linux.nixVersions.nix_2_21 \
  -- build .#default \
  --no-write-lock-file \
  --override-input crane ../.. \
  --store $(pwd)/alt-store
```

**Expected behavior**

The flake should build fine with both Nix 2.20 and 2.21

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by ivan-tkatchev** [comment]
Extremely important issue, currently in Nix 2.21 chroot stores do not work at all.

**Comment by Ericson2314** [comment]
Sounds like #9852

**Comment by edolstra** [comment]
I can confirm that this is fixed by #10345.
